Gems/Terrain/Code/Source/TerrainRenderer/TerrainMeshManager.cpp

```diff
@@ -175,6 +213,66 @@ namespace Terrain
 
     }
 
+    AZ::RHI::StreamBufferView TerrainMeshManager::CreateStreamBufferView(AZ::Data::Instance<AZ::RPI::Buffer>& buffer)
+    {
+        return
+        {
+            *buffer->GetRHIBuffer(),
+            0,
+            aznumeric_cast<uint32_t>(buffer->GetBufferSize()),
+            AZ::RHI::GetFormatSize(buffer->GetBufferViewDescriptor().m_elementFormat)
+        };
+    }
+
+    void TerrainMeshManager::BuildDrawPacket(StackSectorData& sector)
+    {
+        AZ::RHI::DrawPacketBuilder drawPacketBuilder;
+        drawPacketBuilder.Begin(nullptr);
+        drawPacketBuilder.SetDrawArguments(AZ::RHI::DrawIndexed(1, 0, 0, m_indexBuffer->GetBufferViewDescriptor().m_elementCount, 0));
+        drawPacketBuilder.SetIndexBufferView(m_indexBufferView);
+        drawPacketBuilder.AddShaderResourceGroup(sector.m_srg->GetRHIShaderResourceGroup());
+        drawPacketBuilder.AddShaderResourceGroup(m_materialInstance->GetRHIShaderResourceGroup());
+
+        sector.m_perDrawSrgs.clear();
+
+        for (CachedDrawData& drawData : m_cachedDrawData)
+        {
+            AZ::Data::Instance<AZ::RPI::Shader>& shader = drawData.m_shader;
+
+            AZ::RHI::DrawPacketBuilder::DrawRequest drawRequest;
+            drawRequest.m_listTag = drawData.m_drawListTag;
+            drawRequest.m_pipelineState = drawData.m_pipelineState;
+            drawRequest.m_streamBufferViews = sector.m_streamBufferViews;
+            drawRequest.m_stencilRef = AZ::Render::StencilRefs::UseDiffuseGIPass | AZ::Render::StencilRefs::UseIBLSpecularPass;
+
+            const AZ::RPI::ShaderVariantId finalVariantId = drawData.m_shaderOptions.GetShaderVariantId();
+            const AZ::RPI::ShaderVariant& variant = shader->GetVariant(finalVariantId);
+
+            auto drawSrgLayout = shader->GetAsset()->GetDrawSrgLayout(shader->GetSupervariantIndex());
```

Comment:
The variant could be tricky because `GetVariant()` returns a reference - It allows for copies, so maybe worth trying, feels a little dirty though. DrawSrgLayout should be easy to cache.

Reply:
looks like caching both of those things works. 👍